sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncCacheNonBlocking.java

```diff
@@ -0,0 +1,206 @@
+import java.util.function.Function;
+
+/**
+ *
+ * @param <TKey>
+ * @param <TValue>
+ *
+ * This is a thread safe AsyncCache that allows refreshing values in the background.
+ * The benefits of AsyncCacheNonBlocking over AsyncCache is it keeps stale values until the refresh is completed.
+ * AsyncCache removes values causing it to block all requests until the refresh is complete.
+ * 1. For example 1 replica moved out of the 4 replicas available. 3 replicas could still be processing requests.
+ *    The request going to the 1 stale replica would be retried.
+ * 2. AsyncCacheNonBlocking updates the value in the cache rather than recreating it on each refresh. This will help reduce object creation.
+ */
+public class AsyncCacheNonBlocking<TKey, TValue> {
+    private final static Logger logger = LoggerFactory.getLogger(AsyncCacheNonBlocking.class);
+    private final ConcurrentHashMap<TKey, AsyncLazyWithRefresh<TValue>> values;
+
+    public AsyncCacheNonBlocking() {
+        this.values = new ConcurrentHashMap<>();
+    }
+
+    private Boolean removeNotFoundFromCacheException(CosmosException e) {
+        if (Exceptions.isNotFound(e)) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     *
+     * <p>
+     * If another initialization function is already running, new initialization function will not be started.
+     * The result will be result of currently running initialization function.
+     * </p>
+     *
+     * <p>
+     * If previous initialization function is successfully completed it will return the value. It is possible this
+     * value is stale and will only be updated after the force refresh task is complete.
+     * Force refresh is true:
+     *  If the key does not exist: It will create and await the new task
+     *  If the key exists and the current task is still running: It will return the existing task
+     *  If the key exists and the current task is already done: It will start a new task to get the updated values.
+     *     Once the refresh task is complete it will be returned to caller.
+     *     If it is a success the value in the cache will be updated. If the refresh task throws an exception the key will be removed from the cache.
+     * </p>
+     *
+     * <p>
+     * If previous initialization function failed - new one will be launched.
+     * </p>
+     *
+     * @param key Key for which to get a value.
+     * @param singleValueInitFunc Initialization function.
+     * @param forceRefresh Force refresh for refreshing the cache
+     * @return Cached value or value returned by initialization function.
+     */
+    public Mono<TValue> getAsync(
+        TKey key,
+        Function<TValue, Mono<TValue>> singleValueInitFunc,
+        Function<TValue, Boolean> forceRefresh) {
+
+        AsyncLazyWithRefresh<TValue> initialLazyValue = values.get(key);
+        if (initialLazyValue != null) {
+            if (logger.isDebugEnabled()) {
```

Comment:
not needed - actually bad for perf. Only useful when any calculation for logging isn't lazy

Reply:
Removed it.